src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.NetStandard.cs

```diff
@@ -0,0 +1,113 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Security.Cryptography
+{
+    internal sealed partial class SP800108HmacCounterKdfImplementationCng : ISP800108HmacCounterKdfImplementation
+    {
+        internal unsafe SP800108HmacCounterKdfImplementationCng(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithm)
+        {
+            Debug.Assert(hashAlgorithm.Name is not null);
+
+            scoped ReadOnlySpan<byte> symmetricKeyMaterial;
+            scoped Span<byte> clearSpan = default;
+            int symmetricKeyMaterialLength;
+            int hashAlgorithmBlockSize = GetHashBlockSize(hashAlgorithm.Name);
+
+            if (key.Length > hashAlgorithmBlockSize)
+            {
+                byte[] keyArray = new byte[key.Length];
+
+                fixed (byte* pKeyArray = keyArray)
+                {
+                    key.CopyTo(keyArray);
+                    clearSpan = HashOneShot(hashAlgorithm, keyArray);
+                    CryptographicOperations.ZeroMemory(keyArray);
+                }
+
+                symmetricKeyMaterial = clearSpan;
+                symmetricKeyMaterialLength = symmetricKeyMaterial.Length;
+            }
+            else if (!key.IsEmpty)
+            {
+                symmetricKeyMaterial = key;
+                symmetricKeyMaterialLength = key.Length;
+            }
+            else
+            {
+                // CNG requires a non-null pointer even when the length is zero.
+                symmetricKeyMaterial = stackalloc byte[] { 0 };
+                symmetricKeyMaterialLength = 0;
+            }
+
+            try
+            {
+                fixed (byte* pSymmetricKeyMaterial = symmetricKeyMaterial)
+                {
+                    _keyHandle = CreateSymmetricKey(pSymmetricKeyMaterial, symmetricKeyMaterialLength);
+                }
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(clearSpan);
+            }
+
+            _hashAlgorithm = hashAlgorithm;
+        }
+
+        // For .NET Standard / .NET Framework, provide a byte overload so that we don't go from array->span->array
+        // when we need to adjust keys that are too large.
+        internal unsafe SP800108HmacCounterKdfImplementationCng(byte[] key, HashAlgorithmName hashAlgorithm)
+        {
+            Debug.Assert(hashAlgorithm.Name is not null);
+
+            scoped ReadOnlySpan<byte> symmetricKeyMaterial;
+            scoped Span<byte> clearSpan = default;
+            int symmetricKeyMaterialLength;
+            int hashAlgorithmBlockSize = GetHashBlockSize(hashAlgorithm.Name);
+
+            if (key.Length > hashAlgorithmBlockSize)
+            {
+                clearSpan = HashOneShot(hashAlgorithm, key);
```

Comment:
The only difference between the `byte[]` and `ReadOnlySpan<byte>` versions is the HashOneShot method call.  Can we collapse these into one big body via something like

```
_keyHandle = OpenKeyHandle(key, hashAlgorithm);

...

private static SafeSomethingHandle OpenKeyHandle(byte[] key, HashAlgorithmName hashAlgorithmName)
{
    int hashAlgorithmBlockSize = GetHashBlockSize(hashAlgorithm.Name);

    if (key.Length > hashAlgorithmBlockSize)
    {
        ...
        return OpenKeyHandleCore(hashAlgorithm, hashedKey);
    }

    return OpenKeyHandleCore(hashAlgorithm, key.AsSpan());
}

private static SafeSomethingHandle OpenKeyHandle(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithmName)
{
    int hashAlgorithmBlockSize = GetHashBlockSize(hashAlgorithm.Name);

    if (key.Length > hashAlgorithmBlockSize)
    {
        ...
        return OpenKeyHandleCore(hashAlgorithm, hashedKey);
    }

    return OpenKeyHandleCore(hashAlgorithm, key);
}
```

Then the Core just gets to funnel down the "don't pass null", and is common for both paths.

Reply:
I tried this originally and it doesn't really improve much. In addition to the hashing, the clearing of the hashed key makes it difficult to mix and match `byte[]`, `Span<byte>` and `ReadOnlySpan<byte>`. As I worked on refactoring and fixing it, I pretty much just ended up right where I was again.